src/wallet/rpcwallet.cpp

```diff
@@ -2575,6 +2577,234 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+            if (totalFee <= 0)
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid totalFee (cannot be <= 0)");
+            else if (totalFee > maxTxFee)
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid totalFee (higher than maxTxFee)");
+        }
+    }
+
+    // calculate the old fee and fee-rate
+    CAmount nDebit = wtx.GetDebit(ISMINE_SPENDABLE);
+    CAmount nOldFee = (wtx.IsFromMe(ISMINE_SPENDABLE) ? nDebit - wtx.GetValueOut() : 0);
+    int txSize = (int)GetVirtualTransactionSize((CTransaction)wtx);
+    CFeeRate nOldFeeRate(nOldFee, txSize);
+
+    // use the user-defined payTxFee if possible, otherwise use smartfee / fallbackfee
+    CAmount nNewFee = 0;
+    CFeeRate nNewFeeRate;
+    if (payTxFee.GetFeePerK() > 0) {
+        nNewFeeRate = CFeeRate(payTxFee.GetFeePerK());
+    }
+    else {
+        int estimateFoundTarget = newConfirmTarget;
+        nNewFeeRate = mempool.estimateSmartFee(newConfirmTarget, &estimateFoundTarget);
+        if (nNewFeeRate.GetFeePerK() == 0) nNewFeeRate = CWallet::fallbackFee;
+    }
+
+    // new fee rate must be at least old rate + minimum relay rate
+    if (nNewFeeRate.GetFeePerK() < nOldFeeRate.GetFeePerK() + ::minRelayTxFee.GetFeePerK()) {
+        nNewFeeRate = CFeeRate(nOldFeeRate.GetFeePerK() + ::minRelayTxFee.GetFeePerK());
+    }
+
+    // signature sizes can vary by a byte, so add 1 for each input when calculating the new fee
+    nNewFee = nNewFeeRate.GetFee(txSize + wtx.vin.size());
+
+    // if user set totalFee, use that instead
+    if (totalFee > 0) {
+        CAmount minTotalFee = nOldFeeRate.GetFee(txSize) + minRelayTxFee.GetFee(txSize);
+        if (totalFee < minTotalFee) {
+            strError = strprintf("Invalid totalFee, must be at least oldFee + relayFee: %s", FormatMoney(minTotalFee));
+            throw JSONRPCError(RPC_INVALID_PARAMETER, strError);
+        }
+        nNewFee = totalFee;
+        nNewFeeRate = CFeeRate(totalFee, txSize);
+    }
+
+    // check that fee rate is higher than mempool's mininum fee
+    // (no point in bumping fee if we know that the new tx won't be accepted to the mempool)
+    // This may occur if the user set TotalFee or paytxfee too low, if fallbackfee is too low, or, perhaps,
+    // in a rare situation where the mempool minimum fee increased significantly since the fee estimation just a
+    // moment earlier. In this case, we report an error to the user, who may use totalFee to make an adjustment.
+    CFeeRate minMempoolFeeRate = mempool.GetMinFee(GetArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 1000000);
+    if (nNewFeeRate.GetFeePerK() < minMempoolFeeRate.GetFeePerK()) {
+        strError = strprintf("New fee rate (%s) is too low to get into the mempool (min rate: %s)", FormatMoney(nNewFeeRate.GetFeePerK()), FormatMoney(minMempoolFeeRate.GetFeePerK()));
+        throw JSONRPCError(RPC_MISC_ERROR, strError);
+    }
+
+    CAmount nDelta = nNewFee - nOldFee;
```

Comment:
There is kind of a weird deal here where nDelta can be <= 0 if totalFee is set, and will be accepted on a race condition with mempool, I would explicitly guard against this case by restricting nDelta > 0.